### PR TITLE
Refactor curl connection for mocking

### PR DESCRIPTION
**Pull request by vhvb1989**
This PR separates the network transfer logic into one component and provides an interface for it.

Then `gmock` is used to mock the behavior of the derived class of that interface to change and simulate the server responses.

fixes: https://github.com/Azure/azure-sdk-for-cpp/issues/870

### CODE REVIEW

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+#include <curl/curl.h>
+#include <list>
+#include <map>
+#include <memory>
+#include <mutex>
+
+#ifdef TESTING_BUILD
+// Define the class name that reads from ConnectionPool private members
+namespace Azure { namespace Core { namespace Test {
+  class TransportAdapter_connectionPoolTest_Test;
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:23]: The name you selected doesn't meet our guidelines.
```suggestion
  class ConnectionPoolTest;
```

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp [hunk 2]
```diff
@@ -0,0 +1,108 @@
...
+   * more than one request. Use this component when connections are not re-used by default.
+   *
+   * This pool offers static methods and it is allocated statically. There can be only one
+   * connection pool per application.
+   */
+  struct CurlConnectionPool
+  {
+#ifdef TESTING_BUILD
+    // Give access to private to this tests class
+    friend class Azure::Core::Test::TransportAdapter_connectionPoolTest_Test;
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:40]:
```suggestion
    friend class Azure::Core::Test::CurlConnectionPoolTest;
```

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp [hunk 3]
```diff
@@ -0,0 +1,108 @@
...
+#endif
+
+    /**
+     * @brief Mutex for accessing connection pool for thread-safe reading and writing.
+     */
+    static std::mutex s_connectionPoolMutex;
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:46]: This is a public member.
```suggestion
    static std::mutex ConnectionPoolMutex;
```

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp [hunk 4]
```diff
@@ -0,0 +1,108 @@
...
+    /**
+     * @brief Keeps an unique key for each host and creates a connection pool for each key.
+     *
+     * @detail This way getting a connection for a specific host can be done in O(1) instead of
+     * looping a single connection list to find the first connection for the required host.
+     *
+     * @remark There might be multiple connections for each host.
+     */
+    static std::map<std::string, std::list<std::unique_ptr<CurlNetworkConnection>>>
+        s_connectionPoolIndex;
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:57]:
```suggestion
        ConnectionPoolIndex;
```

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp [hunk 5]
```diff
@@ -0,0 +1,108 @@
...
+namespace Azure { namespace Core { namespace Http {
+
+  /**
+   * @brief CURL HTTP connection pool makes it possible to re-use one curl connection to perform
+   * more than one request. Use this component when connections are not re-used by default.
+   *
+   * This pool offers static methods and it is allocated statically. There can be only one
+   * connection pool per application.
+   */
+  struct CurlConnectionPool
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:36]:
```suggestion
  class CurlConnectionPool
```

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp (hunk 3 above)
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:46]:
```suggestion
public:
    static std::mutex ConnectionPoolMutex;
```

File: sdk/core/azure-core/test/ut/curl_session.hpp [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+namespace Azure { namespace Core { namespace Test {
+
+  class CurlSession : public ::testing::Test {
+  };
+
+  /**
+   * @brief mock the network connection
+   *
+   */
+  class CurlNetworkConnectionMock : public Azure::Core::Http::CurlNetworkConnection {
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/curl_session.hpp:19]: I suggest we standardize on mock being a prefix.  It reads better when its written "Mock of X" vs "X Mock"

```suggestion
  class MockCurlNetworkConnection : public Azure::Core::Http::CurlNetworkConnection {
```

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp [hunk 7]
```diff
@@ -0,0 +1,108 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
```
Comment (RickWinter) [inline sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:1]: Name of file is missing _

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp (hunk 1 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:23]: It gets generated by gtest.
The name is only `connectionPoolTest`. gtest adds the class name and the label `Test` at the end

File: sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp (hunk 3 above)
Comment (vhvb1989) [reply sdk/core/azure-core/inc/azure/core/http/curl/curl connection_pool.hpp:46]: `s` is for static

File: sdk/core/azure-core/test/ut/curl_session_test.cpp [hunk 8]
```diff
@@ -0,0 +1,44 @@
...
+namespace Azure { namespace Core { namespace Test {
+
+  TEST_F(CurlSession, successCall)
+  {
+    std::string response(
+        "HTTP/1.1 200 Ok\r\nContent-Type: text/html; charset=UTF-8\r\n\r\n{\r\n\"somejson\":45\r}");
+
+    // Can't mock the curMock directly from a unique ptr, heap allocate it first and then make a
+    // unique ptr for it
+    MockCurlNetworkConnection* curlMock = new MockCurlNetworkConnection();
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/curl_session_test.cpp:23]: This line of code leaks.  You need to delete the memory you allocate when you are done with it

File: sdk/core/azure-core/test/ut/transport_adapter.hpp [hunk 9]
```diff
@@ -1,6 +1,14 @@
 // Copyright (c) Microsoft Corporation. All rights reserved.
 // SPDX-License-Identifier: MIT
 
+/**
```
Comment (RickWinter) [inline sdk/core/azure-core/test/ut/transport_adapter.hpp:4]: Consider naming this something specific to tests to make clear this is a test only class

**Review by RickWinter — APPROVED** [review]

File: sdk/core/azure-core/test/ut/curl_session_test.cpp (hunk 8 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/curl_session_test.cpp:23]: gmock is intelligent and tells you if there is anything leak at the end of the test.

On line 31, We create a unique_ptr from this pointer to take care about deleting the memory.

I also add a mock destructor to make sure we are destructing the object :)

File: sdk/core/azure-core/test/ut/transport_adapter.hpp (hunk 9 above)
Comment (vhvb1989) [reply sdk/core/azure-core/test/ut/transport_adapter.hpp:4]: we would then need to rename all the test files, to be consistent.

**Pull request merged**